internal/app/ldap/ldap.go

```diff
@@ -79,54 +96,106 @@ func (l *LdapAuthentication) Authenticate(username string, password string) (int
 	return responmap, nil
 
 }
-func ldapauthentication(username string, password string, cf LDAPconf, l glog.Logger) error {
-	ldapconn, err := ldap.Dial("tcp", fmt.Sprintf("%s:%d", cf.LdapAddr, cf.LdapPort))
+func ldapauthentication(username string, password string, cf LDAPconf, l glog.Logger, conn *ldap.Conn) error {
+	userdn, err := finduserdn(username, cf, l, conn)
+	l.Infof("userdn= %v", userdn)
 	if err != nil {
-		l.Errorf("canan't connect to ldap server")
 		return err
 	}
-	defer ldapconn.Close()
-	var sb strings.Builder
-	sb.WriteString("Cn=")
-	sb.WriteString(username)
-	sb.WriteString(",")
-	sb.WriteString(cf.BaseDN)
-	err = ldapconn.Bind(sb.String(), password)
-	if err != nil {
-		l.Errorf("can't authentication ")
-		return err
+	ok := false
+	for _, dn := range userdn {
+		err = conn.Bind(dn, password)
+		if err == nil {
+			ok = true
+			break
+		}
+	}
+	if ok {
+		return nil
 	}
-	return nil
+	return errors.New("login fails")
+
 }
-func ldapquery(username string, cf LDAPconf, l glog.Logger) (types.User, error) {
-	ldapconn, err := ldap.Dial("tcp", fmt.Sprintf("%s:%d", cf.LdapAddr, cf.LdapPort))
+func ldapquery(username string, cf LDAPconf, l glog.Logger, conn *ldap.Conn) (types.User, error) {
+	err := conn.Bind(cf.LdapFirstBindUsername, cf.LdapFirstBindPassword)
 	if err != nil {
 		return types.User{}, err
 	}
-	defer ldapconn.Close()
-	err = ldapconn.Bind(cf.LdapFirstBindUsername, cf.LdapFirstBindPassword)
+	searchRequest := ldap.NewSearchRequest(
+		cf.BaseDN,
+		ldap.ScopeWholeSubtree, ldap.NeverDerefAliases, 0, 0, false,
+		fmt.Sprintf("(&(objectClass=*)(cn=%s))", username),
+		[]string{"dn", "mail", "memberOf", "description", "displayName", "cn"},
+		nil,
+	)
+	searchresult, err := conn.Search(searchRequest)
 	if err != nil {
+		l.Errorf("ldap query search fail")
+		return types.User{}, err
+	}
+	if len(searchresult.Entries) == 0 {
+		l.Errorf("no user query")
+		err := errors.New("user does not exist ")
 		return types.User{}, err
 	}
+	for _, entries := range searchresult.Entries {
+		if strlib.IsLower(entries.GetAttributeValue("cn")) {
```

Comment:
Why do we need to do this? And it doesn't make sense to introduce a new lib just because you want to do a little thing like check if it's lower or not.